firmware/src/core/utils/MainHelper.cpp

```diff
@@ -0,0 +1,226 @@
+        } else if (buttonId == BUTTON_RIGHT) {
+            Serial.printf("Right button pressed, state=%d\n", state);
+            s_widgetCycleDelayPrev = millis();
+            s_widgetSet->buttonPressed(BUTTON_RIGHT, state);
+        }
+    }
+}
+
+void MainHelper::checkButtons() {
+    ButtonState leftState = buttonLeft.getState();
+    if (leftState != BTN_NOTHING) {
+        buttonPressed(BUTTON_LEFT, leftState);
+    }
+    ButtonState middleState = buttonOK.getState();
+    if (middleState != BTN_NOTHING) {
+        buttonPressed(BUTTON_OK, middleState);
+    }
+    ButtonState rightState = buttonRight.getState();
+    if (rightState != BTN_NOTHING) {
+        buttonPressed(BUTTON_RIGHT, rightState);
+    }
+}
+
+void MainHelper::checkCycleWidgets() {
+    if (s_widgetSet && s_widgetCycleDelay > 0 && (s_widgetCycleDelayPrev == 0 || (millis() - s_widgetCycleDelayPrev) >= s_widgetCycleDelay * 1000)) {
+        s_widgetSet->next();
+        s_widgetCycleDelayPrev = millis();
+    }
+}
+
+// Handle simulated button state
+void MainHelper::handleEndpointButton() {
+    if (s_wifiManager->server->hasArg("name") && s_wifiManager->server->hasArg("state")) {
+        String inButton = s_wifiManager->server->arg("name");
+        String inState = s_wifiManager->server->arg("state");
+        uint8_t buttonId = Utils::stringToButtonId(inButton);
+        ButtonState state = Utils::stringToButtonState(inState);
+
+        if (buttonId != 0 && state != BTN_NOTHING) {
+            buttonPressed(buttonId, state);
+            s_wifiManager->server->send(200, "text/plain", "OK " + inButton + "/" + inState + " -> " + String(buttonId) + "/" + String(state));
+            return;
+        }
+    }
+    s_wifiManager->server->send(500, "text/plain", "ERR");
+}
+
+// Show button web page
+void MainHelper::handleEndpointButtons() {
+    String msg = WEBPORTAL_BUTTONS_PAGE_START1;
+    msg += WEBPORTAL_BUTTONS_STYLE;
+    msg += WEBPORTAL_BUTTONS_PAGE_START2;
+    String buttons[] = {"left", "middle", "right"};
+    String states[] = {"short", "medium", "long"};
+    int numButtons = 3; // number of buttons
+    int numStates = 3; // number of states
+    for (int s = 0; s < numStates; s++) {
+        msg += "<tr>";
+        for (int b = 0; b < numButtons; b++) {
+            msg += "<td><button class='sim' onclick=\"sendReq('" + buttons[b] + "', '" + states[s] + "')\">" + buttons[b] + "<br>" + states[s] + "</button></td>";
+        }
+        msg += "</tr>";
+    }
+    msg += WEBPORTAL_BUTTONS_PAGE_END1;
+    msg += WEBPORTAL_BUTTONS_SCRIPT;
+    msg += WEBPORTAL_BUTTONS_PAGE_END2;
+    s_wifiManager->server->send(200, "text/html", msg);
+}
+
+void MainHelper::setupWebPortalEndpoints() {
+    // To simulate button presses call e.g. http://<ip>/button?name=right&state=short
+    s_wifiManager->server->on("/button", handleEndpointButton);
+    s_wifiManager->server->on("/buttons", handleEndpointButtons);
+}
+
+void MainHelper::showWelcome() {
```

Comment:
Be careful of creating too many 'helper' and 'util' classes.
In general, those should be fairly clean and only contain stuff that _really_ does not belong anywhere.

For example, this function looks like it could be a widget,
something like the 'WelcomeWidget' or the likes?

Reply:
`MainHelper` is supposed to include all the minor stuff from `main.cpp`, because `main.cpp` became pretty big and cluttered.

Now, only the high level stuff remains in `main.cpp` and the implementation details are in `MainHelper`.

Showing the Welcome message was done in `main` before, therefore I moved it to `MainHelper`.
A widget would certainly be possible, but that's a bit more complicated because `WifiWidget` also writes to the screens at the same time.

I would suggest refactoring this later because creating a `WelcomeWidget` is not really in the scope of this PR.